orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,706 @@
+        self.titleArea.addWidget(self.off_label)
+        self.off_label.hide()
+
+        self.titleArea.addStretch()
+
+        if toggle_enabled:
+            self.toggle_module_switch = QCheckBox()
+            self.toggle_module_switch.stateChanged.connect(self.on_toggle)
+            switch_icon_on_resource = _i('on_button.png')
+            switch_icon_off_resource = _i('off_button.png')
+            style_sheet = '''
+            QCheckBox::indicator {
+                width: 23px;
+                height: 23px;
+            }
+            QCheckBox::indicator:checked {
+                image: url(%s);
+            }
+            QCheckBox::indicator:unchecked {
+                image: url(%s);
+            }
+            ''' % (switch_icon_off_resource, switch_icon_on_resource)
+            self.toggle_module_switch.setStyleSheet(style_sheet)
+
+            self.titleArea.addWidget(self.toggle_module_switch)
+
+        # Root.
+        self.rootArea = QVBoxLayout()
+        self.rootArea.setContentsMargins(0, 0, 0, 0)
+        self.rootArea.setSpacing(0)
+        self.setLayout(self.rootArea)
+        self.rootArea.addWidget(title_holder)
+
+    def add_widget(self, widget):
+        # For appending widgets to the root layout of the widget.
+        self.rootArea.addWidget(widget)
+
+    def notify_on_change(self):
+        # Emits signals corresponding to the changes done.
+        self.change_signal.emit()
+
+    def on_toggle(self):
+        # Activated when the widget is enabled/disabled.
+        self.enabled = not self.enabled
+        if self.enabled:
+            self.off_label.hide()
+            self.title_label.setStyleSheet('color: #000000;')
+        else:
+            self.off_label.show()
+            self.title_label.setStyleSheet('color: #B0B0B0;')
+        self.change_signal.emit()
+
+    def restore_data(self, data):
+        # Restores the widget state from the input data.
+        raise NotImplementedError
+
+    def export_data(self):
+        # Export the settings for this module instance.
+        return NotImplementedError
+
+    @staticmethod
+    def get_pp_settings():
+        # Returns the dict representation of this portion of a pre-processor.
+        return NotImplementedError
+
+
+class TokenizerModule(PreprocessorModule):
+    DEFAULT_SETTINGS = {
+        'is_enabled': True,
+        'method': 0,
+    }
+
+    NLTKTokenizer, TwitterTokenizer = 0, 1
+    tokenizer_values = {
+        NLTKTokenizer: False,
+        TwitterTokenizer: True
+    }
+    tokenizer_names = {
+        NLTKTokenizer: 'NLTK tokenizer',
+        TwitterTokenizer: 'Twitter tokenizer',
+    }
+
+    tokenizer_method = 0
+
+    def __init__(self, data):
+        if not data:
+            data = self.DEFAULT_SETTINGS
+        PreprocessorModule.__init__(self, 'Tokenizer', False,
+                                    data.get('is_enabled'))
+
+        self.contents = QWidget()
```

Comment:
The following four lines are the same for all widgets. Could this be abstracted into the `PreprocessorModule` class? Then we could some method like `self.add_to_content_area` that would add buttons, checkboxes, etc directly into the `self.contentArea`.